include/pmacc/particles/algorithm/ForEach.hpp

```diff
@@ -22,157 +22,447 @@
-                         *                must fulfill the interface pmacc::functor::Interface<F, 1u, void>
-                         * @param mapper functor to map a block to a supercell
-                         * @param pb particles species box
-                         */
-                        template<typename T_Acc, typename T_Functor, typename T_Mapping, typename T_ParBox>
-                        DINLINE void operator()(
-                            T_Acc const& acc,
-                            T_Functor functor,
-                            T_Mapping const mapper,
-                            T_ParBox pb) const
-                        {
-                            using SuperCellSize = typename T_ParBox::FrameType::SuperCellSize;
-                            constexpr uint32_t dim = SuperCellSize::dim;
-                            constexpr uint32_t frameSize = pmacc::math::CT::volume<SuperCellSize>::type::value;
-                            constexpr uint32_t numWorkers = T_numWorkers;
+                return pb.getFirstFrame(superCellIdx);
+            }
 
-                            uint32_t const workerIdx = cupla::threadIdx(acc).x;
+            template<typename T_ParBox>
+            static DINLINE auto next(T_ParBox& pb, typename T_ParBox::FramePtr const& framePtr)
+            {
+                return pb.getNextFrame(framePtr);
+            }
 
-                            DataSpace<dim> const superCellIdx(
-                                mapper.getSuperCellIndex(DataSpace<dim>(cupla::blockIdx(acc))));
+            template<typename T_ForEach>
+            static DINLINE auto createCtx(T_ForEach& forEachParticleInFrame)
+            {
+                return lockstep::makeVar<uint32_t>(forEachParticleInFrame, 0u);
+            }
+        };
 
-                            auto const& superCell = pb.getSuperCell(superCellIdx);
-                            uint32_t const numPartcilesInSupercell = superCell.getNumParticles();
+        //! Policy to reverse iterate over particles/frames of a supercell.
+        struct Reverse
+        {
+            template<typename T_ParBox>
+            static DINLINE auto getFirst(T_ParBox& pb, DataSpace<T_ParBox::Dim> const& superCellIdx)
+            {
+                return pb.getLastFrame(superCellIdx);
+            }
 
+            template<typename T_ParBox>
+            static DINLINE auto next(T_ParBox& pb, typename T_ParBox::FramePtr const& framePtr)
+            {
+                return pb.getPreviousFrame(framePtr);
+            }
 
-                            // end kernel if we have no particles
-                            if(numPartcilesInSupercell == 0)
-                                return;
+            template<typename T_ForEach>
+            static DINLINE auto createCtx(T_ForEach& forEachParticleInFrame)
+            {
+                return lockstep::makeVar<lcellId_t>(forEachParticleInFrame, lcellId_t(0u));
```

Comment:
It is different from `Forward<>::createCtx()` on purpose? Here you use `lcellId_t`, and there `uint32_t`.

Reply:
Yes depending on the iteration order the variable is used differently. `lcellId_t` is enough to store the number of particles in a frame.